# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
# Commit: 1913c332695e
# Full Hash: 1913c332695e2776ee7f034a3a528ad465d6b13a
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - Set main thread as current TaskQueue when calling into libwebrtc so that video calling works. r=bwc
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D104721
# ==============================================================================

diff -r 7d270e96dfbf -r 1913c332695e dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
--- a/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Mon Feb 08 18:15:07 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Wed Feb 10 09:58:11 2021 +0100
@@ -431,6 +431,7 @@
 
 void WebrtcVideoConduit::DeleteSendStream() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   if (mSendStream) {
@@ -441,6 +442,7 @@
 
 MediaConduitErrorCode WebrtcVideoConduit::CreateSendStream() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   nsAutoString codecName;
@@ -471,6 +473,7 @@
 
 void WebrtcVideoConduit::DeleteRecvStream() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   if (mRecvStream) {
@@ -481,6 +484,7 @@
 
 MediaConduitErrorCode WebrtcVideoConduit::CreateRecvStream() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   mRecvStreamConfig.decoders.clear();
@@ -530,6 +534,7 @@
 MediaConduitErrorCode WebrtcVideoConduit::ConfigureSendMediaCodec(
     const VideoCodecConfig* codecConfig, const RtpRtcpConfig& aRtpRtcpConfig) {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   MutexAutoLock lock(mMutex);
 
   mUpdateSendResolution = true;
@@ -726,6 +731,7 @@
 
 bool WebrtcVideoConduit::SetRemoteSSRCLocked(uint32_t ssrc, uint32_t rtxSsrc) {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   if (mRecvStreamConfig.rtp.remote_ssrc == ssrc &&
@@ -814,6 +820,7 @@
 Maybe<webrtc::VideoReceiveStream::Stats> WebrtcVideoConduit::GetReceiverStats()
     const {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   if (!mRecvStream) {
     return Nothing();
   }
@@ -823,6 +830,7 @@
 Maybe<webrtc::VideoSendStream::Stats> WebrtcVideoConduit::GetSenderStats()
     const {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   if (!mSendStream) {
     return Nothing();
   }
@@ -831,12 +839,14 @@
 
 webrtc::Call::Stats WebrtcVideoConduit::GetCallStats() const {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   return mCall->Call()->GetStats();
 }
 
 void WebrtcVideoConduit::GetRtpSources(
     nsTArray<dom::RTCRtpSourceEntry>& outSources) {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   std::vector<webrtc::RtpSource> sources = mRecvStream->GetSources();
   outSources.Clear();
   for (const auto& source : sources) {
@@ -953,6 +963,7 @@
 
 void WebrtcVideoConduit::DeleteStreams() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
 
   using namespace Telemetry;
   if (mSendBitrate.NumDataValues() > 0) {
@@ -1129,6 +1140,7 @@
        (mRecvStreamConfig.rtp.ulpfec_payload_type != ulpfec_payload_type ||
         mRecvStreamConfig.rtp.red_payload_type != red_payload_type))) {
     MutexAutoLock lock(mMutex);
+    auto current = TaskQueueWrapper::MainAsCurrent();
 
     condError = StopReceivingLocked();
     if (condError != kMediaConduitNoError) {
@@ -1362,11 +1374,30 @@
                                                         int len) {
   ASSERT_ON_THREAD(mStsThread);
 
-  // Bug 1499796 - we need to get passed the time the packet was received
+  using PacketPromise =
+      MozPromise<webrtc::PacketReceiver::DeliveryStatus, bool, true>;
+
+  auto syncPromise =
+      InvokeAsync(GetMainThreadSerialEventTarget(), __func__, [&] {
+        auto current = TaskQueueWrapper::MainAsCurrent();
+        if (!mCall->Call()) {
+          return PacketPromise::CreateAndResolve(
+              webrtc::PacketReceiver::DELIVERY_PACKET_ERROR, __func__);
+        }
+        // Bug 1499796 - we need to get passed the time the
+        // packet was received
+        webrtc::PacketReceiver::DeliveryStatus status =
+            mCall->Call()->Receiver()->DeliverPacket(
+                webrtc::MediaType::VIDEO,
+                rtc::CopyOnWriteBuffer(static_cast<const uint8_t*>(data), len),
+                -1);
+        return PacketPromise::CreateAndResolve(status, __func__);
+      });
+
   webrtc::PacketReceiver::DeliveryStatus status =
-      mCall->Call()->Receiver()->DeliverPacket(
-          webrtc::MediaType::VIDEO,
-          rtc::CopyOnWriteBuffer(static_cast<const uint8_t*>(data), len), -1);
+      media::Await(GetMediaThreadPool(MediaThreadType::WEBRTC_DECODER),
+                   syncPromise)
+          .ResolveValue();
 
   if (status != webrtc::PacketReceiver::DELIVERY_OK) {
     CSFLogError(LOGTAG, "%s DeliverPacket Failed, %d", __FUNCTION__, status);
@@ -1514,6 +1545,7 @@
 
 MediaConduitErrorCode WebrtcVideoConduit::StopTransmittingLocked() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   if (mEngineTransmitting) {
@@ -1530,6 +1562,7 @@
 
 MediaConduitErrorCode WebrtcVideoConduit::StartTransmittingLocked() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   if (mEngineTransmitting) {
@@ -1557,6 +1590,7 @@
 
 MediaConduitErrorCode WebrtcVideoConduit::StopReceivingLocked() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   // Are we receiving already? If so, stop receiving and playout
@@ -1574,6 +1608,7 @@
 
 MediaConduitErrorCode WebrtcVideoConduit::StartReceivingLocked() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
   mMutex.AssertCurrentThreadOwns();
 
   if (mEngineReceiving) {
@@ -1766,6 +1801,7 @@
 
 void WebrtcVideoConduit::CollectTelemetryData() {
   MOZ_ASSERT(NS_IsMainThread());
+  auto current = TaskQueueWrapper::MainAsCurrent();
 
   if (mEngineTransmitting) {
     webrtc::VideoSendStream::Stats stats = mSendStream->GetStats();
